src/main/java/com/redhat/devtools/intellij/quarkus/lsp4ij/operations/diagnostics/LSPDiagnosticAnnotator.java

```diff
@@ -57,77 +54,53 @@ public LSPVirtualFileWrapper collectInformation(@NotNull PsiFile file, @NotNull
-    public void apply(@NotNull PsiFile file, LSPVirtualFileWrapper wrapper, @NotNull AnnotationHolder holder) {
-        if (wrapper.isDiagnosticDirty()) {
-            // LSP Diagnostics has been published for the given file, create for each diagnostic an Intellij Annotation
-            createAnnotations(wrapper, holder);
-            // FIXME : remove this following code, code action request should be called when cursor change or when error annotation is hovered!
-            // Associate for each annotation some quick fixes coming from the code actions request
-            List<Annotation>  annotations = wrapper.getAnnotations();
-            if(annotations != null) {
-                for (Annotation annotation : annotations) {
-                    wrapper.updateQuickFixAt(annotation.getStartOffset());
-                }
-            }
-        } else  {
-            // Update annotation holder with existing annotations
-            // This case comes from:
-            // - when an external component from Intellij Quarkus refresh the validation with DaemonCodeAnalyzer.getInstance(module.getProject()).restart(psiFile);
-            // - when quick fix are updated from LSP code actions.
-         updateAnnotations(wrapper.getAnnotations(), holder);
-        }
-    }
+    public void apply(@NotNull PsiFile file, LSPVirtualFileWrapper editorWrapper, @NotNull AnnotationHolder holder) {
+        // Get current LSP diagnostics of the current file
+        LSPVirtualFileWrapper fileWrapper = LSPVirtualFileWrapper.getLSPVirtualFileWrapper(file.getVirtualFile());
+        final Collection<LSPDiagnosticsForServer> diagnosticsPerServer = fileWrapper.getAllDiagnostics();
+        Document document = LSPIJUtils.getDocument(file.getVirtualFile());
 
-    private void createAnnotations(LSPVirtualFileWrapper wrapper, AnnotationHolder holder) {
-        final List<Diagnostic> diagnostics = wrapper.getAllDiagnostics();
-        VirtualFile virtualFile = wrapper.getFile();
-        Document document = LSPIJUtils.getDocument(virtualFile);
-
-        List<Annotation> annotations = new ArrayList<>();
-        diagnostics.forEach(diagnostic -> {
-            Annotation annotation = createAnnotation(document, diagnostic, holder);
-            if (annotation != null) {
-                if (diagnostic.getTags() != null && diagnostic.getTags().contains(DiagnosticTag.Deprecated)) {
-                    annotation.setHighlightType(ProblemHighlightType.LIKE_DEPRECATED);
-                }
-                annotations.add(annotation);
+        // Loop for language server which report diagnostics for the given file
+        diagnosticsPerServer.forEach(ds -> {
+            boolean codeActionsLoading = false;
+            // Loop for LSP diagnostics to transform it to Intellij annotation.
+            for (Diagnostic diagnostic : ds.getDiagnostics()) {
+                codeActionsLoading = codeActionsLoading | createAnnotation(diagnostic, document, ds, holder);

```

Comment:
Not sure if you have a test scenario for this, but I'm using a language server that does not support codeactions, and is triggering a couple exceptions that are difficult to track down. 
But it seems like it starts from `createAnnotation` returning true here.

Reply:
After modifying some logic here, it's still happening, so it may be that the `codeAction` is called from somewhere else without checking if the language server supports it?

Reply:
Indeed I don't take care of that, plesase review my PR at https://github.com/redhat-developer/intellij-quarkus/pull/827